File: problems/518.零钱兑换-ii.py
#
# @lc app=leetcode.cn id=518 lang=python3
#
# [518] 零钱兑换 II
#

# @lc code=start

class Solution:
    def change(self, amount: int, coins: list) -> int:
        # 这个题目用一维的分析好像不行，例如 1 2 组成 3；这样出来 1 + dp[2] 和 2 + dp[1] 是一种方法，就重复了
        # 但是这个题目和 01 背包问题非常相似， dp[i][j] 表示使用前 i 个凑成面值为 j 的组合数
        # 做空间复杂度的优化
        dp = [[0 for _ in range(1+amount)]for _ in range(2)]
        # 处理第一行
        for j in range(1+amount):
            if 0 == j % coins[0]:  # 经过题目测试样例测试， amount 为 0 的时候返回 1
                dp[0][j] = 1

        for i in range(1, len(coins)):
            for j in range(1+amount):
                if j-coins[i] < 0:
                    dp[i & 1][j] = dp[1 - i & 1][j]
                else:
                    # 注意，这个货币可能可以使用很多次，所以需要循环
                    dp[i & 1][j] = dp[1 - i & 1][j]
                    k = j
                    while True:
                        k -= coins[i]
                        if k < 0:
                            break
                        dp[i & 1][j] += dp[1 - i & 1][k]
        return dp[(len(coins)-1) & 1][amount]


# @lc code=end

File: problems/518.零钱兑换-ii.cpp
/*
 * @lc app=leetcode.cn id=518 lang=cpp
 *
 * [518] 零钱兑换 II
 */

// @lc code=start

#include <bits/stdc++.h>


class Solution
{

public:

    int change(int amount, std::vector<int> &coins)
    {
        int n = coins.size();

        // 经典完全背包问题，dp[i][j] 表示使用前 i 个凑成面额 j 的组合数（i 从 0 开始）
        // 二维 vector 可以这样初始化，第一个参数是第一维的大小，第二个参数就是初始化第二维的 vector
        std::vector<std::vector<int>> dp(n, std::vector<int>(1 + amount));

        // 初始化第一行
        for (int j = 0; j <= amount; ++j) dp[0][j] = (0 == j % coins[0]) ? 1 : 0;

        // 继续迭代
        for (int i = 1; i < n; ++i)
        {
            for (int j = 0; j <= amount; ++j)
            {
                // 可以不使用第 i 个
                dp[i][j] = dp[i - 1][j];

                // 如果要使用需要符合面值条件
                if (j >= coins[i]) dp[i][j] += dp[i][j - coins[i]];
            }
        }

        return dp[n - 1][amount];
    }
};

// @lc code=end
